Use API package to create orders by batch

rodelgc opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?

### Changes proposed in this Pull Request:

Currently in `trunk`, the e2e test `order-status-filters.test.js` creates test orders through the UI. This PR replaces these setup steps to use the ["Batch Create Orders" API](https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-orders) to improve the speed of this test.

Based on the latest CI runs in `trunk` and this branch, notice that we've halved the execution time of this test:
<img width="634" alt="Screen Shot 2021-07-01 at 1 19 31 PM" src="https://user-images.githubusercontent.com/4509348/124069839-725f3800-da6f-11eb-92d0-9ea9a29a0bc8.png">
<img width="634" alt="Screen Shot 2021-07-01 at 1 20 40 PM" src="https://user-images.githubusercontent.com/4509348/124069856-7a1edc80-da6f-11eb-954d-8a62ac177f8a.png">


Closes #29797.

### How to test the changes in this Pull Request:

1. Launch the e2e test environment
2. Run the `activate-and-setup` tests: `npx wc-e2e test:e2e ./tests/e2e/specs/activate-and-setup/setup-onboarding.js` 
3. Run the test that this PR changed.
    Headless mode: `npx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/order-status-filters.test.js`
    UI  mode: `npx wc-e2e test:e2e-dev ./tests/e2e/specs/wp-admin/order-status-filters.test.js`
4. They should both pass
<img width="457" alt="Screen Shot 2021-06-30 at 7 01 26 PM" src="https://user-images.githubusercontent.com/4509348/123949825-9bcb8580-d9d5-11eb-9c90-bda0bf9ef941.png">

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

In `core-tests`:
> Updated the `order-status-filters.test.js` e2e test to use the API in creating orders

In `e2e-utils`:
> New function to create orders by batch using the orders API

Commit:
Used API package to create orders by batch

Commit:
Added changelog entries

zhongruige reviewed (commented):
These ran great for me locally, thanks for adding in these change @rodelgc!

Looks like there's just a minor merge conflict on the `CHANGLOG.md` file we just want to clear up before we can merge.

Once batch support is added with https://github.com/woocommerce/woocommerce/issues/29570 we can switch over to using that as part of the API package. But I think this implementation works great for now 👍

rodelgc commented:
Thanks @zhongruige. I'll sort out the merge conflict some time today or tomorrow.

rrennick commented on tests/e2e/utils/src/components.js:
```diff
@@ -352,6 +352,31 @@ const createSimpleOrder = async ( orderStatus = 'Pending payment' ) => {
+	const path = '/wc/v3/orders/batch';
+
+	// Create an order per status
+	const orders = statuses.map((s) => {
+		return {
+			...defaultOrder,
+			status: s
+		};
+	});
```
This would be more readable if the variable was `status`:

```
const orders = statuses.map((status) => {
		return {
			...defaultOrder,
			status
		};
	});
```

rodelgc replied:
Thanks @rrennick. Fixed in https://github.com/woocommerce/woocommerce/pull/30184/commits/b116756b0967dde7fd61de6edbdcf2790a4b056c

rrennick commented on tests/e2e/utils/src/components.js:
```diff
@@ -352,6 +352,31 @@ const createSimpleOrder = async ( orderStatus = 'Pending payment' ) => {
+ * 
+ * @param statuses Array of order statuses
+ */
+const batchCreateOrders = async (statuses) => {
```
Can we move this function to the `withRestApi` flow?

rodelgc replied:
Done. Fixed in https://github.com/woocommerce/woocommerce/pull/30184/commits/16083c257478115f613f2e28366a8d16d2f3525f

rrennick commented on tests/e2e/utils/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## Added
 
 - Factories for variable product, variation, and grouped product
+- New function to create orders by batch using the orders API
```
All utilities functions are listed in the [package readme](https://github.com/woocommerce/woocommerce/tree/trunk/tests/e2e/utils#readme). Whenever we add a new utility the PR should add the function to the list.

rodelgc replied:
Done. Fixed in https://github.com/woocommerce/woocommerce/pull/30184/commits/16083c257478115f613f2e28366a8d16d2f3525f

rrennick reviewed (changes requested):
@rodelgc Nice work on this. I've asked for a few minor changes. Once those are done this should be good to merge.

Commit:
Renamed 'status' variable

Commit:
Move utility function to withRestApi

Commit:
Add new util function to the readme

Commit:
Merge branch 'trunk' into add/e2e-merchant-batch-create-orders

rodelgc commented on tests/e2e/utils/src/flows/with-rest-api.js:
```diff
@@ -152,5 +153,31 @@ export const withRestApi = {
+		const payload = { create: orders };
+		const { statusCode } = await client.post(path, payload);
+
+		/* eslint-disable jest/no-standalone-expect */
```
@zhongruige @rrennick I had to put this otherwise I get an eslint error when I try to push my changes. I know that we discussed this in Slack and based on my understanding, we will fix this in the coming sprints. But for now, is it ok for us to add `eslint-disable` comments like this, or is there a more elegant way workaround? Thanks!

rrennick replied:
@rodelgc You can remove that line. The jest linting rules were removed in https://github.com/woocommerce/woocommerce/pull/30206.

rrennick replied:
@rodelgc I misread your question. In our other files we have disabled the jest linting at the top of the file. The reason you get that error running the tests is that the function is called inside `beforeAll` vs inside `it`.

rodelgc commented:
> @rodelgc Nice work on this. I've asked for a few minor changes. Once those are done this should be good to merge.

Thanks @rrennick I've already addressed the issues you raised, and it's ready for another review.

I left a question [above](https://github.com/woocommerce/woocommerce/pull/30184#discussion_r679087294). Let me know if the approach I did there is fine.

rrennick commented on tests/e2e/utils/src/flows/with-rest-api.js:
```diff
@@ -152,5 +153,31 @@ export const withRestApi = {
+	 * @param statuses Array of order statuses
+	 */
+	batchCreateOrders : async (statuses) => {
+		const defaultOrder = config.get('orders.basicPaidOrder');
```
This utility function will be more useful to package users if `defaultOrder` is read in the test function and passed as a parameter. That way other users can pass whatever order data they want to the function.

rrennick replied:
@rodelgc I forgot to ping you on the last two comments on this PR.

rodelgc replied:
Hi @rrennick no worries. I already started working on this earlier this week. 

> That way other users can pass whatever order data they want to the function.

Makes sense. I'll have the orders created inside the test instead, and then pass it as an array (of orders) to `batchCreateOrders()`

rodelgc replied:
> I'll have the orders created inside the test instead,

I believe the best way to do this is to use the utility function `createSimpleOrder()` found in `tests/e2e/utils/src/components.js`. But `createSimpleOrder()`  is using the UI to create the order. So I'm tempted to refactor `createSimpleOrder()` in such a way that it would use the API instead.

However, I saw from 116-gh-woocommerce/woocommerce-quality#issuecomment-900589554 that @zhongruige is already updating some of the tests to use the API. I'll double check with him to see if he already included `createSimpleOrder()` in his updates so that we avoid conflicts.

rodelgc replied:
Sorry, never mind my comment above 😅 

I just realized that `createSimpleOrder()` actually creates the order in the store, not the order as a JS object. Got confused for a minute there.

rodelgc replied:
Done. Fixed in https://github.com/woocommerce/woocommerce/pull/30184/commits/71fe61a594514fb45371a0a661bcaff226a06eec

rrennick commented on tests/e2e/utils/src/components.js:
```diff
@@ -521,5 +521,5 @@ export {
 	clickUpdateOrder,
 	deleteAllEmailLogs,
-	deleteAllShippingZones,
+	deleteAllShippingZones
```
You editor may be automatically removing these. The reason we add the optional comma is that when adding imports or exports in another PR we don't have a line in the change set to add the comma.

rodelgc replied:
You're right. I found out my prettier was pointing to a different config file. This is fixed now in https://github.com/woocommerce/woocommerce/pull/30184/commits/261f41525d5eb64865ab46e1718bc4b7de449fcd

rodelgc commented:
Hi @rrennick I just saw your newer comments just now ([this](https://github.com/woocommerce/woocommerce/pull/30184#discussion_r682900511) and [this](https://github.com/woocommerce/woocommerce/pull/30184#discussion_r682902210)). I'll push a fix for them either today or on Monday. Thanks for the review!

Commit:
Replaced trailing comma

Commit:
batchCreateOrder now accepts array of orders as parameter

Commit:
Merge branch 'trunk' into add/e2e-merchant-batch-create-orders

rodelgc commented:
The E2E tests passed but the `Run unit tests on PR / PHP 7.4 WP nightly` job is failing, which is a known issue in WP as discussed here - p1629211878056700-slack-C01DT6U03HC

This PR is ready for another round of review :)

rrennick reviewed (approved):
Nice work on this @rodelgc . It tested great locally and looks like it has quite a nice performance improvement.

```
Test Suites: 1 skipped, 38 passed, 38 of 39 total
Tests:       2 skipped, 164 passed, 166 total
Snapshots:   0 total
Time:        725.013s
Ran all test suites.
Jest exit code: 0
```
